File: typescript/src/routing/IRouteStore.ts
/**
 * Route store abstraction + in-memory default.
 * SPDX-License-Identifier: MIT
 */

import { RouteEntry, isRouteExpired } from "../models/index.js";

export interface IRouteStore {
  get(destinationUhid: string): Promise<RouteEntry | null>;
  getAll(): Promise<RouteEntry[]>;
  save(route: RouteEntry): Promise<void>;
  remove(destinationUhid: string): Promise<void>;
  pruneExpired(): Promise<number>;
}

export class InMemoryRouteStore implements IRouteStore {
  private readonly routes = new Map<string, RouteEntry>();

  async get(destinationUhid: string): Promise<RouteEntry | null> {
    return this.routes.get(destinationUhid) ?? null;
  }

  async getAll(): Promise<RouteEntry[]> {
    return Array.from(this.routes.values());
  }

  async save(route: RouteEntry): Promise<void> {
    this.routes.set(route.destinationUhid, route);
  }

  async remove(destinationUhid: string): Promise<void> {
    this.routes.delete(destinationUhid);
  }

  async pruneExpired(): Promise<number> {
    let pruned = 0;
    const now = new Date();
    for (const [k, v] of Array.from(this.routes.entries())) {
      if (isRouteExpired(v, now)) {
        this.routes.delete(k);
        pruned++;
      }
    }
    return pruned;
  }
}

File: src/AetherMesh.Core/Routing/IRouteStore.cs
// SPDX-License-Identifier: MIT

using AetherMesh.Models;

namespace AetherMesh.Routing;

/// <summary>
/// Persistent backing store for the routing table. The default implementation
/// (<see cref="InMemoryRouteStore"/>) is process-local; production hosts typically
/// substitute a SQLite- or file-backed implementation so routes survive restarts.
/// </summary>
public interface IRouteStore
{
    /// <summary>Returns the route to <paramref name="destinationUhid"/>, or null if none is stored.</summary>
    Task<RouteEntry?> GetAsync(string destinationUhid, CancellationToken cancellationToken = default);

    /// <summary>Returns every route currently stored, including expired ones (pruning is the caller's responsibility).</summary>
    Task<IReadOnlyList<RouteEntry>> GetAllAsync(CancellationToken cancellationToken = default);

    /// <summary>Inserts or replaces the route for <see cref="RouteEntry.DestinationUhid"/>.</summary>
    Task SaveAsync(RouteEntry route, CancellationToken cancellationToken = default);

    /// <summary>Removes the route for <paramref name="destinationUhid"/>, if present.</summary>
    Task RemoveAsync(string destinationUhid, CancellationToken cancellationToken = default);

    /// <summary>Removes every route whose <see cref="RouteEntry.ExpiresAt"/> is in the past.</summary>
    Task<int> PruneExpiredAsync(CancellationToken cancellationToken = default);
}
